File: HERO C#/CANifier Demo/Tasks/TaskMainLoop.cs
/**
 * If a USB Game controller (Logitech wireless, Xbox 360, similar) is connected use that for control
 * If no gamepad is present or user flipped D-X Switch to disable, disable the robot.
 * Additionally the shoulder buttons switch the Arm and Wheel between closed loop and open loop modes.
 **/
using CTRE.Phoenix.Tasking;
using CTRE.Phoenix;
using Platform;

public class TaskMainLoop : ILoopable
{
    public void OnLoop()
    {
        bool gamepadOk = false;

        /* keep feeding watchdog to enable motors */
        if (Hardware.gamepad.GetConnectionStatus() == UsbDeviceConnection.Connected)
        {
            /* keep talons enabled */
            CTRE.Phoenix.Watchdog.Feed();
            /* take note if gamepad is present */
            gamepadOk = true;
        }

        if (gamepadOk)
        {
            Platform.Schedulers.PeriodicTasks.Start(Platform.Tasks.taskPWMmotorController);
        }
        else
        {
            Platform.Schedulers.PeriodicTasks.Stop(Platform.Tasks.taskPWMmotorController);
        }


        if (!gamepadOk || Hardware.gamepad.GetButton(6))
        {
            /* no gamepad present OR user is holding R shoulder btn. Just roll thru color wheel*/
            Platform.Schedulers.PeriodicTasks.Start(Platform.Tasks.taskAnimateLEDStrip);
            Platform.Schedulers.PeriodicTasks.Stop(Platform.Tasks.taskDirectControlArm);
            Platform.Schedulers.PeriodicTasks.Stop(Platform.Tasks.taskLIDAR_ControlLEDStrip);
        }
        else if (Hardware.gamepad.GetButton(5))
        {
            /* let user control LED with sticks */
            Platform.Schedulers.PeriodicTasks.Stop(Platform.Tasks.taskAnimateLEDStrip);
            Platform.Schedulers.PeriodicTasks.Start(Platform.Tasks.taskDirectControlArm);
            Platform.Schedulers.PeriodicTasks.Stop(Platform.Tasks.taskLIDAR_ControlLEDStrip);

            Platform.Schedulers.PeriodicTasks.Start(Platform.Tasks.taskMeasurePulseSensors);
        }
        else if (Hardware.gamepad.GetButton(7))
        {
            Platform.Schedulers.PeriodicTasks.Stop(Platform.Tasks.taskAnimateLEDStrip);
            Platform.Schedulers.PeriodicTasks.Stop(Platform.Tasks.taskDirectControlArm);
            Platform.Schedulers.PeriodicTasks.Start(Platform.Tasks.taskLIDAR_ControlLEDStrip);

            Platform.Schedulers.PeriodicTasks.Start(Platform.Tasks.taskMeasurePulseSensors);
        }
    }

    public void OnStart()
    {
        /* default to LED strip animation */
        Schedulers.PeriodicTasks.Start(Platform.Tasks.taskAnimateLEDStrip);
        Schedulers.PeriodicTasks.Stop(Platform.Tasks.taskDirectControlArm);
        Schedulers.PeriodicTasks.Stop(Platform.Tasks.taskLIDAR_ControlLEDStrip);
    }
    public void OnStop() { }
    public bool IsDone() { return false; }
}

File: C++ General/CANifier Demo/src/main/cpp/Tasks/TaskMainLoop.cpp
#include "Tasks/TaskMainLoop.h"
#include "Platform/Platform.h"

/* Destructor */
TaskMainLoop::~TaskMainLoop() { }

/* ILoopable */
void TaskMainLoop::OnStart() {
	/* default to LED strip animation */
	Schedulers::PeriodicTasks->Start(Tasks::taskAnimateLEDStrip);
	Schedulers::PeriodicTasks->Stop(Tasks::taskDirectControlArm);
	Schedulers::PeriodicTasks->Stop(Tasks::taskLIDARControlLEDStrip);
}
void TaskMainLoop::OnStop() {
}
bool TaskMainLoop::IsDone() {
	return false;
}
void TaskMainLoop::OnLoop() {
	bool controllerState = false;

	/* No method of determining if controller is connected */
	if (controllerState == false) {
		Schedulers::PeriodicTasks->Stop(Tasks::taskPWMmotorController);
	} else {
		/* Just for testing, can be enabled by user by changing controllerState to true */
		Schedulers::PeriodicTasks->Start(Tasks::taskPWMmotorController);
	}

	if (Hardware::gamepad->GetRawButton(6)) {
		/* no gamepad present OR user is holding R shoulder btn. Just roll thru color wheel*/
		Schedulers::PeriodicTasks->Start(Tasks::taskAnimateLEDStrip);
		Schedulers::PeriodicTasks->Stop(Tasks::taskDirectControlArm);
		Schedulers::PeriodicTasks->Stop(Tasks::taskLIDARControlLEDStrip);
	} else if (Hardware::gamepad->GetRawButton(5)) {
		/* let user control LED with sticks */
		Schedulers::PeriodicTasks->Stop(Tasks::taskAnimateLEDStrip);
		Schedulers::PeriodicTasks->Start(Tasks::taskDirectControlArm);
		Schedulers::PeriodicTasks->Stop(Tasks::taskLIDARControlLEDStrip);

		Schedulers::PeriodicTasks->Start(Tasks::taskMeasurePulseSensors);

	} else if (Hardware::gamepad->GetRawButton(7)) {
		Schedulers::PeriodicTasks->Stop(Tasks::taskAnimateLEDStrip);
		Schedulers::PeriodicTasks->Stop(Tasks::taskDirectControlArm);
		Schedulers::PeriodicTasks->Start(Tasks::taskLIDARControlLEDStrip);

		Schedulers::PeriodicTasks->Start(Tasks::taskMeasurePulseSensors);

	}
}

